sqld/src/replication/primary/logger.rs

```diff
@@ -720,28 +743,89 @@ pub struct ReplicationLogger {
 impl ReplicationLogger {
     pub fn open(db_path: &Path, max_log_size: u64) -> anyhow::Result<Self> {
         let log_path = db_path.join("wallog");
+        let data_path = db_path.join("data");
+
+        let fresh = !log_path.exists();
+
         let file = OpenOptions::new()
             .create(true)
             .write(true)
             .read(true)
             .open(log_path)?;
+
         let max_log_frame_count = max_log_size * 1_000_000 / LogFile::FRAME_SIZE as u64;
         let log_file = LogFile::new(file, max_log_frame_count)?;
+        let header = log_file.header();
+
+        let should_recover = if header.version < 2 || header.sqld_version() != Version::current() {
+            tracing::info!("replication log version not compatible with current sqld version, recovering from database file.");
+            true
+        } else if fresh && data_path.exists() {
+            tracing::info!("replication log not found, recovering from database file.");
+            true
+        } else {
+            false
+        };
+
+        if should_recover {
+            Self::recover(log_file, data_path)
+        } else {
+            Self::from_log_file(db_path.to_path_buf(), log_file)
+        }
+    }
 
-        let header = log_file.header;
+    fn from_log_file(db_path: PathBuf, log_file: LogFile) -> anyhow::Result<Self> {
+        let header = log_file.header();
         let generation_start_frame_no = header.start_frame_no + header.frame_count;
 
         let (new_frame_notifier, _) = watch::channel(generation_start_frame_no);
 
         Ok(Self {
             generation: Generation::new(generation_start_frame_no),
-            compactor: LogCompactor::new(db_path, log_file.header.db_id)?,
+            compactor: LogCompactor::new(&db_path, log_file.header.db_id)?,
             log_file: RwLock::new(log_file),
-            db_path: db_path.to_owned(),
+            db_path,
             new_frame_notifier,
         })
     }
 
+    fn recover(log_file: LogFile, mut data_path: PathBuf) -> anyhow::Result<Self> {
+        tracing::info!("Replication log not found, or corrupted: recovering from database file");
+        let mut log_file = log_file.reset()?;
```

Comment:
for the sake of debugging sessions and disaster recovery, I suggest only truncating the file after we successfully checkpointed the database - it simply leaves us more context if something fails.

Reply:
not sure how that would help, but it wouldn't hurt either